pkg/controller/certificaterequests/vault/vault_test.go

```diff
@@ -0,0 +1,326 @@
+					`Warning ErrorSigning Vault failed to sign certificate: failed to sign`,
+				},
+				CheckFn: testcr.MustNoResponse,
+			},
+			fakeVault:   fakevault.New().WithSign(nil, nil, errors.New("failed to sign")),
+			expectedErr: false,
+		},
+		"a client with a app role secret referenced with role but failed to sign should report fail": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerVault(v1alpha1.VaultIssuer{
+					Auth: v1alpha1.VaultAuth{
+						AppRole: v1alpha1.VaultAppRole{
+							RoleId: "my-role-id",
+							SecretRef: v1alpha1.SecretKeySelector{
+								LocalObjectReference: v1alpha1.LocalObjectReference{
+									"role-secret",
+								},
+								Key: "my-role-key",
+							},
+						},
+					},
+				}),
+			),
+			certificateRequest: testCR,
+			builder: &testpkg.Builder{
+				KubeObjects:        []runtime.Object{roleSecret},
+				CertManagerObjects: []runtime.Object{},
+				ExpectedEvents: []string{
+					`Warning ErrorSigning Vault failed to sign certificate: failed to sign`,
+				},
+				CheckFn: testcr.MustNoResponse,
+			},
+			fakeVault:   fakevault.New().WithSign(nil, nil, errors.New("failed to sign")),
+			expectedErr: false,
+		},
+		"a client with a token secret referenced with token and signs should return certificate": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerVault(v1alpha1.VaultIssuer{
+					Auth: v1alpha1.VaultAuth{
+						TokenSecretRef: v1alpha1.SecretKeySelector{
+							Key: "my-token-key",
+							LocalObjectReference: v1alpha1.LocalObjectReference{
+								"token-secret",
+							},
+						},
+					},
+				}),
+			),
+			certificateRequest: testCR,
+			builder: &testpkg.Builder{
+				KubeObjects:        []runtime.Object{tokenSecret},
+				CertManagerObjects: []runtime.Object{},
+				ExpectedEvents:     []string{},
+				CheckFn:            testcr.NoPrivateKeyFieldsSetCheck(rsaPEMCert),
+			},
+			fakeVault:   fakevault.New().WithSign(rsaPEMCert, rsaPEMCert, nil),
+			expectedErr: false,
+		},
+		"a client with a app role secret referenced with role should return certificate": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerVault(v1alpha1.VaultIssuer{
+					Auth: v1alpha1.VaultAuth{
+						AppRole: v1alpha1.VaultAppRole{
+							RoleId: "my-role-id",
+							SecretRef: v1alpha1.SecretKeySelector{
+								LocalObjectReference: v1alpha1.LocalObjectReference{
+									"role-secret",
+								},
+								Key: "my-role-key",
+							},
+						},
+					},
+				}),
+			),
+			certificateRequest: testCR,
+			builder: &testpkg.Builder{
+				KubeObjects:        []runtime.Object{roleSecret},
+				CertManagerObjects: []runtime.Object{},
+				ExpectedEvents:     []string{},
+				CheckFn:            testcr.NoPrivateKeyFieldsSetCheck(rsaPEMCert),
+			},
+			fakeVault:   fakevault.New().WithSign(rsaPEMCert, rsaPEMCert, nil),
+			expectedErr: false,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			runTest(t, test)
+		})
+	}
```

Comment:
I'm still concerned here that we aren't validating the output state of the CertificateRequest in any of these tests (i.e. checking that the 'status' conditions have been set appropriately).

We sort of indirectly test it for *some cases* via the `ExpectedEvents` structure, however all these tests would be far simpler if we dropped use of the CheckFn altogether and simply validated outputs in terms of ExpectedActions (plus it's trivial to encode the logic that the current default CheckFns do into actions.. simply check the status field of the CR).